source.go

```diff
@@ -53,46 +68,69 @@ func (s *Source) Configure(ctx context.Context, cfg map[string]string) error {
+	sdk.Logger(ctx).Info().Msg("Opening Source...")
+	s.server = NewServer(ctx)
+	err := s.runServer()
+	if err != nil {
+		return err
+	}
 	return nil
 }
 
 func (s *Source) Read(ctx context.Context) (sdk.Record, error) {
-	// Read returns a new Record and is supposed to block until there is either
-	// a new record or the context gets cancelled. It can also return the error
-	// ErrBackoffRetry to signal to the SDK it should call Read again with a
-	// backoff retry.
-	// If Read receives a cancelled context or the context is cancelled while
-	// Read is running it must stop retrieving new records from the source
-	// system and start returning records that have already been buffered. If
-	// there are no buffered records left Read must return the context error to
-	// signal a graceful stop. If Read returns ErrBackoffRetry while the context
-	// is cancelled it will also signal that there are no records left and Read
-	// won't be called again.
-	// After Read returns an error the function won't be called again (except if
-	// the error is ErrBackoffRetry, as mentioned above).
-	// Read can be called concurrently with Ack.
-	return sdk.Record{}, nil
+	select {
+	case <-ctx.Done():
+		return sdk.Record{}, ctx.Err()
+	case record, ok := <-s.server.recordCh:
+		if !ok {
+			return sdk.Record{}, fmt.Errorf("record channel is closed")
+		}
+		return record, nil
+	}
 }
 
 func (s *Source) Ack(ctx context.Context, position sdk.Position) error {
-	// Ack signals to the implementation that the record with the supplied
-	// position was successfully processed. This method might be called after
-	// the context of Read is already cancelled, since there might be
-	// outstanding acks that need to be delivered. When Teardown is called it is
-	// guaranteed there won't be any more calls to Ack.
-	// Ack can be called concurrently with Read.
+	select {
+	case s.server.ackCh <- position:
+		// worked fine!
+	default:
+		return fmt.Errorf("ack channel is closed")
+	}
+	// wait until ack is sent into the stream
+	sent := <-s.server.ackSent
+	if !sent {
+		return fmt.Errorf("failed to send acknowledgment %q to client", position)
+	}
+
 	return nil
 }
 
 func (s *Source) Teardown(ctx context.Context) error {
-	// Teardown signals to the plugin that there will be no more calls to any
-	// other function. After Teardown returns, the plugin should be ready for a
-	// graceful shutdown.
+	if s.grpcSrv != nil {
+		s.grpcSrv.Stop()
+		s.wg.Wait()
+	}
+	return nil
+}
+
+func (s *Source) runServer() error {
+	// s.listener can be set for test purposes
+	if s.listener == nil {
+		lis, err := net.Listen("tcp", s.config.URL)
+		if err != nil {
+			return fmt.Errorf("failed to listen: %w", err)
+		}
+		s.listener = lis
+	}
+	s.grpcSrv = grpc.NewServer()
+	pb.RegisterStreamServiceServer(s.grpcSrv, s.server)
+
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+		if err := s.grpcSrv.Serve(s.listener); err != nil {
+			log.Fatalf("failed to serve: %v", err)
```

Comment:
Whoah, `log.Fatalf` :flushed: This calls `os.Exit(1)` and kills the process, only use this in throwaway code 🙃

Reply:
oh no that's definitely a forgotten test log 🥲

Reply:
Hmm, idk. I didn't dive deeper to know what exactly can provoke an error here. We could try to be on the safe side and propagate it to the teardown function, since that's where we wait for this goroutine to stop. So instead of having a wait group we could have a channel that returns an error or gets closed when the server stops.